plugins/s3/src/plugin.cpp

```diff
@@ -0,0 +1,307 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/location.hpp>
+#include <tenzir/plugin.hpp>
+
+#include <arrow/filesystem/filesystem.h>
+#include <arrow/filesystem/s3fs.h>
+#include <arrow/filesystem/type_fwd.h>
+#include <arrow/io/api.h>
+#include <arrow/util/uri.h>
+#include <fmt/core.h>
+
+namespace tenzir::plugins::s3 {
+
+namespace {
+
+struct s3_args {
+  bool anonymous;
+  located<std::string> uri;
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, s3_args& x) -> bool {
+    return f.object(x).pretty_name("s3_args").fields(
+      f.field("anonymous", x.anonymous), f.field("uri", x.uri));
+  }
+};
+
+auto get_options(const s3_args& args) -> arrow::fs::S3Options {
+  auto opts = arrow::fs::S3Options::Anonymous();
+  if (not args.anonymous) {
+    auto opts_from_uri = arrow::fs::S3Options::FromUri(args.uri.inner);
+    if (not opts_from_uri.ok()) {
+      opts = arrow::fs::S3Options::Defaults();
+    } else {
+      opts = *opts_from_uri;
+    }
+  }
+  return opts;
+}
+
+// We use 2^20 for the upper bound of a chunk size, which exactly matches the
+// upper limit defined by execution nodes for transporting events.
+// TODO: Get the backpressure-adjusted value at runtime from the execution node.
+constexpr size_t max_chunk_size = 1 << 20;
+} // namespace
+
+class s3_loader final : public plugin_loader {
+public:
+  s3_loader() = default;
+
+  s3_loader(s3_args args) : args_{std::move(args)} {
+  }
+  auto instantiate(operator_control_plane& ctrl) const
+    -> std::optional<generator<chunk_ptr>> override {
+    return
+      [](s3_args args, operator_control_plane& ctrl) -> generator<chunk_ptr> {
+        auto initialized = arrow::fs::EnsureS3Initialized();
+        if (not initialized.ok()) {
+          diagnostic::error("failed to initialize Arrow S3 functionality: {}",
+                            initialized.ToString())
+            .emit(ctrl.diagnostics());
+          co_return;
+        }
+        auto guard = caf::detail::make_scope_guard([&ctrl]() {
+          auto finalized = arrow::fs::FinalizeS3();
+          if (not finalized.ok()) {
+            diagnostic::error("failed to close Arrow S3 filesystem: {}",
+                              finalized.ToString())
+              .emit(ctrl.diagnostics());
```

Comment:
```
error: !! filesystem_error: failed to initialize Arrow S3 functionality: Invalid: Attempt to initialize S3 after it has been finalized
```

I can reproduce this error reliably if I let a managed pipeline writing to or reading from S3 run through, and then try to start a second managed pipeline that does the same. I think we don't want to close the S3 filesystem here yet, but rather do it at the level of the S3 plugin and in its destructor. Similarly, we probably want to initialize it only once in the plugin's `initialize(...)` function.

I assume that the same applies to the GCS plugin, but have not yet verified that.

Reply:
I can see this being necessary. Have we thought about what this means wrt. to multiple S3 pipelines using different authentication data? If we initialize S3 only once, it's available only with one global configuration and not per pipeline.

Reply:
It's not a problem if you pass the parameters to the URI, e.g., `load s3 s3://bucket/file.json?endpoint_override=http://localhost:9000&secret_key=secret`.

As a user I would also not expect environment variables to be read more than once, so that's also not a problem.

Reply:
Yeah, I crossed  _another_ Arrow stumbling block with this one.

I aligned S3 initialization time with the connector initialization time.

Reply:
Note for the future: putting credentials in the pipeline definition is as bad as putting them on the command line. We need to think about how to factor this, e.g., with a config file as stop gap and potentially things like a secret store later on.

Reply:
I think that's usually not a problem. You choose one S3-compatible store, and then put the credentials in the config file or the env variables that the AWS SDK expects. Doing overrides on a per-pipeline basis is not the common case.

Reply:
I think Arrow already handles those stop gaps, by using [AWS's credentials provider chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials-chain.html).

We are trying to get options from the URI, with a fallback to default credentials, from our side and Arrow's side:
https://github.com/apache/arrow/blob/1adc745f67791dde1d9e2574b9de52a8d00ca7e8/cpp/src/arrow/filesystem/s3fs.cc#L364

Reply:
> I assume that the same applies to the GCS plugin, but have not yet verified that.
The Arrow GCS Filesystem does not have such initialization; you can `Make()` multiple GCS filesystems, and destruction will not affect further initialization.

This one-time initialization thing is actually more like an AWS SDK peculiarity than Arrow's.

Reply:
> Note for the future: putting credentials in the pipeline definition is as bad as putting them on the command line. We need to think about how to factor this, e.g., with a config file as stop gap and potentially things like a secret store later on.

My tested workaround for now is to specify the endpoint on the URL but the credentials in environment variables.

Reply:
> I aligned S3 initialization time with the connector initialization time.

This doesn't actually work. Arrow's S3 deinitialization gets called before the `plugin` destructor, which leads to segmentation faults.

@dominiklohmann I can maybe implement a `plugins::deinitialize` loop in Tenzir's `main` function, or just (de-)initialize S3 there since `s3` is going to be a builtin anyway.

Reply:
I think another easy way to fix this is to manually destroy all plugins at the end of the main loop.

Right before the `return EXIT_SUCCESS` at the very end of `main`, destroy all plugins by calling `plugins::get_mutable().clear()`.

Reply:
`clear()`ed.

Reply:
This `clear()`ing seems to cause a segmentation fault in the macOS CI step. @dominiklohmann can you reproduce/debug this?

Reply:
Replaced with `plugin` deinitialization. Thank you @dominiklohmann